### PR TITLE
ENH Adds label_smoothing to cross entropy loss

### LINKED ISSUE #4281: Update reference scripts to use the "Batteries Included" utils
## 🚀 Feature
As part of the "Batteries Included" initiative (#3911) we are adding a number of new utils that can be used to produce SOTA results. Once those utils are landed, we should update our reference scripts to use them. 

More specifically we need to:
- [ ] Update the reference scripts to use PyTorch's new [linear warmup](https://github.com/pytorch/pytorch/pull/60836) scheduler.
  - The reference scripts should be updated to let users define the `warmup_method` (if any), the `warmup_iters` and the `warmup_factor`. The warmup scheduler should be chained with other existing schedulers.
  - The Object Detection recipe needs to be BC compatible and replace our [custom linear warmup](https://github.com/pytorch/vision/blob/dc42f933f3343c76727dbfba6e4242f1bcb8e1a0/references/detection/engine.py#L20-L24) approach with the one from PyTorch.
  - Similarly the Video Classification recipe needs also to be BC compatible and replace our [customer scheduler](https://github.com/pytorch/vision/blob/master/references/video_classification/scheduler.py#L5).
  - All other recipes (Classification and Segmentation), should be updated to optionally use warm up. The addition should be BC compatible and turned off by default.
- [x] Update the Classification reference to use PyTorch's new label smoothing implementation.
  - We will support the case of label smoothing by using the upcoming `label_smoothing` argument of CrossEntropyLoss (https://github.com/pytorch/pytorch/pull/63122). 
  - A new `label_smoothing` argument should be introduced on the [script](https://github.com/pytorch/vision/blob/c2ab0c59f42babf9ad01aa616cd8a901daac86dd/references/classification/train.py#L165). To ensure BC, its default value should be 0.0 (no label smoothing).

### LINKED ISSUE #7455: [PyTorch][Feature Request] Label Smoothing for CrossEntropyLoss
Hi, guys. The type `torch.LongTensor` of target will hinder the implementation like some methods in [reference](https://arxiv.org/abs/1512.00567). So is there a possible to add a `Arg: label_smoothing` for `torch.nn.CrossEntropyLoss()`, or maybe simply add the docs to show how to convert the `target` into `one-hot vector` to work with `torch.nn.CrossEntropyLoss()` together, or any other simple ways? Thanks.

cc @ezyang @gchanan @zou3519 @bdhirsh @albanD @mruberry

### PR DESCRIPTION
**Pull request by thomasjpfan**
Fixes https://github.com/pytorch/pytorch/issues/7455

Partially resolves pytorch/vision#4281

cc @albanD @mruberry @jbschlosser

### CODE REVIEW

File: aten/src/ATen/native/LossNLL.cpp [hunk 1]
```diff
@@ -497,23 +497,78 @@ Tensor cross_entropy_loss_prob_target(
...
         "Expected floating point type for target with class probabilities, got ", target.scalar_type());
     TORCH_CHECK(ignore_index < 0, "ignore_index is not supported for floating point target");
+    TORCH_CHECK(label_smoothing == 0.0, "target must be specified as class indices to use label_smoothing")
 
     // See [Note: hacky wrapper removal for optional tensor]
     c10::MaybeOwned<Tensor> weight_maybe_owned = at::borrow_from_optional_tensor(weight);
     const Tensor& weight_ = *weight_maybe_owned;
     ret = cross_entropy_loss_prob_target(self, target, weight_, reduction);
+  } else if (0.0 < label_smoothing && label_smoothing < 1.0) {
+    TORCH_CHECK(ignore_index < 0, "ignore_index is not supported to use label_smoothing");
```
Comment (thomasjpfan) [inline aten/src/ATen/native/LossNLL.cpp:566]: Technically, `ignore_index` can be supported. I disabled it to be consistent with soft targets.

File: aten/src/ATen/native/LossNLL.cpp [hunk 2]
```diff
@@ -497,23 +497,78 @@ Tensor cross_entropy_loss_prob_target(
...
+
+    // See [Note: hacky wrapper removal for optional tensor]
+    c10::MaybeOwned<Tensor> weight_maybe_owned = at::borrow_from_optional_tensor(weight);
+    const Tensor& weight_ = *weight_maybe_owned;
+    ret = cross_entropy_loss_label_smoothing(self, target, weight_, reduction, label_smoothing);
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:571]: AFAICT it should be possible to just call `at::nll_loss_nd` normally and do a linear combination of that and some other label smoothing math for the case where `label_smoothing > 0.0` (see https://github.com/pytorch/pytorch/issues/7455#issuecomment-893652776). Does the weight handling complicate this?

File: aten/src/ATen/native/LossNLL.cpp [hunk 3]
```diff
@@ -497,23 +497,78 @@ Tensor cross_entropy_loss_prob_target(
...
+    Tensor smooth_loss;
+
+    if (weight.defined()) {
+      // Expand weight to the correct number of dims for broadcasting with input / target
+      auto weight_broadcast_shape = SmallBuffer<int64_t, 5>(input.dim());
+      std::fill(weight_broadcast_shape.begin(), weight_broadcast_shape.end(), 1);
+      weight_broadcast_shape[1] = weight.size(0);
+      Tensor weight_ = weight.view(weight_broadcast_shape);
+
+      smooth_loss = -at::sum(input * weight_, 1);
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:522]: Sorry if I misled you during our convo, but I think weight handling for the class index target case should stay consistent with what it's doing now. I could be wrong, but I think it'd be unexpected that setting label smoothing changes how weights are applied.

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -1103,6 +1103,9 @@ class probabilities only when a single class label per minibatch item is too res
             and :attr:`reduce` are in the process of being deprecated, and in
             the meantime, specifying either of those two args will override
             :attr:`reduction`. Default: ``'mean'``
+        label_smoothing (float, optional): A float in [0.0, 1.0]. Specifies the amount of smoothing
+            when computing the loss, where 0.0 means no smoothing. This feature is supported
+            when targets are specified as class indices.
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:1108]: I suggest adding a reference to the paper here to be clear about what type of label smoothing is going on. The original paper that introduced it as a regularization method is [here](https://arxiv.org/pdf/1512.00567.pdf).

File: test/test_nn.py [hunk 5]
```diff
@@ -16897,6 +16897,49 @@ def test_cross_entropy_loss_one_hot_target(self, device):
...
+        with self.assertRaisesRegex(RuntimeError,
+                                    r'target must be specified as class indices to use label_smoothing'):
+            loss(input, bad_target)
+
+        loss = nn.CrossEntropyLoss(label_smoothing=0.3, ignore_index=1)
+        with self.assertRaisesRegex(RuntimeError, r'ignore_index is not supported to use label_smoothing'):
+            loss(input, target)
+
+
+    def test_cross_entropy_label_smoothing_consistent_index_target_and_probs(self, device):
```
Comment (jbschlosser) [inline test/test_nn.py:16916]: I think it'd be good to also have a set of tests that compare outputs when `label_smoothing > 0.0` is set to a reference of manually applied label smoothing + NLL loss. For reference, the formula for how to smooth labels from the paper is here:
```
class_probs = (1 - eps) * class_probs + eps / num_classes
```

**Review by jbschlosser — COMMENTED** [review]
Thanks for tackling this! Added some comments below. To summarize:
* I think we want the end result to handle both `ignore_index` and class prob targets
* Can the implementation be simplified by doing label smoothing as additional math outside of the call to `at::nll_loss_nd` and doing a linear combination?

File: aten/src/ATen/native/LossNLL.cpp (hunk 2 above)
Comment (thomasjpfan) [reply aten/src/ATen/native/LossNLL.cpp]: The weights and `ignore_index` does complicate things. I updated the PR to try to make it as simple as possible.

**Comment by thomasjpfan** [comment]
I updated the PR with:

1. `label_smoothing` for soft targets.
2. Use the exact definition of label smoothing from the original paper.
3. Adds `ignore_index` which ignores targets == `ignore_index`.
4. The semantics of `reduction=='mean'` with weights is consistent with the hard target definition.

File: aten/src/ATen/native/LossNLL.cpp [hunk 6]
```diff
@@ -475,34 +484,92 @@ Tensor cross_entropy_loss_prob_target(
...
+      Tensor weight_ = weight.view(weight_broadcast_shape);
+
+      input *= weight_;
+    }
+    smooth_loss = -at::sum(input, 1);
+
+    if (ignore_index >= 0) {
+      auto ignore_mask = target.eq(ignore_index);
+      smooth_loss.masked_fill_(ignore_mask, 0.0);
+    }
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:540]:
```suggestion
    smooth_loss.index_put_({target == ignore_index}, 0.0);
```

File: aten/src/ATen/native/LossNLL.cpp [hunk 7]
```diff
@@ -475,34 +484,92 @@ Tensor cross_entropy_loss_prob_target(
...
+
+    Tensor ret;
+    switch (reduction) {
+      case Reduction::Mean:
+        if (weight.defined()) {
+          ret = at::sum(smooth_loss);
+          // loss is normalized by the weights to be consistent with nll_loss_nd
+          ret /= weight.gather(0, target.flatten()).sum();
+        } else {
+          ret = at::mean(smooth_loss);
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:549]: nit: `smooth_loss.mean()`

File: aten/src/ATen/native/LossNLL.cpp [hunk 8]
```diff
@@ -475,34 +484,92 @@ Tensor cross_entropy_loss_prob_target(
...
+        }
+        break;
+      case Reduction::Sum:
+        ret = at::sum(smooth_loss);
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:553]: nit: `smooth_loss.sum()`

File: aten/src/ATen/native/LossNLL.cpp [hunk 9]
```diff
@@ -475,34 +484,92 @@ Tensor cross_entropy_loss_prob_target(
...
+    if (ignore_index >= 0) {
+      auto ignore_mask = target.eq(ignore_index);
+      smooth_loss.masked_fill_(ignore_mask, 0.0);
+    }
+
+    Tensor ret;
+    switch (reduction) {
+      case Reduction::Mean:
+        if (weight.defined()) {
+          ret = at::sum(smooth_loss);
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp]: nit: `smooth_loss.sum()`

File: aten/src/ATen/native/LossNLL.cpp [hunk 10]
```diff
@@ -466,6 +467,15 @@ Tensor cross_entropy_loss_prob_target(
...
   auto input = at::log_softmax(self, 1, self.scalar_type());
+  Tensor target;
+
+  if (label_smoothing != 0.0) {
+    TORCH_CHECK(0.0 < label_smoothing && label_smoothing <= 1.0, "label_smoothing must be between 0.0 and 1.0. Got: ", label_smoothing);
+    target = target_ * (1 - label_smoothing) + label_smoothing / n_classes;
+  } else{
+    target = target_;
+  }
+
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:478]: nice, I like this better :) it's clearer that the target is being smoothed

File: torch/testing/_internal/common_nn.py [hunk 11]
```diff
@@ -4101,20 +4106,59 @@ def cross_entropy_loss_prob_target_reference(input, target, weight=None, reducti
...
+
+    smooth_loss = -torch.sum(input, 1)
+
+    if ignore_index >= 0:
+        ignore_mask = target == ignore_index
+        smooth_loss.masked_fill_(ignore_mask, 0.0)
+
+    if reduction == 'mean':
+        if weight is not None:
+            ret = torch.sum(smooth_loss) / weight.gather(0, target.flatten()).sum()
```
Comment (jbschlosser) [inline torch/testing/_internal/common_nn.py:4137]: Possibly add a comment here explaining the denominator again. If we ever fix #61309, we'll want to update all these places

File: torch/testing/_internal/common_nn.py [hunk 12]
```diff
@@ -4101,20 +4106,59 @@ def cross_entropy_loss_prob_target_reference(input, target, weight=None, reducti
...
+        else:
+            ret = torch.mean(smooth_loss)
+    elif reduction == 'sum':
+        ret = torch.sum(smooth_loss)
+    else:
+        ret = smooth_loss
+
+    return (1 - label_smoothing) * nllloss + ret * (label_smoothing / C)
```
Comment (jbschlosser) [inline torch/testing/_internal/common_nn.py:4145]: formula looks good!

File: torch/testing/_internal/common_nn.py [hunk 13]
```diff
@@ -4874,6 +4918,141 @@ def padding3d_circular(input, pad):
         desc='4d_prob_target',
         check_bfloat16=False,
     ),
+    dict(
```
Comment (jbschlosser) [inline torch/testing/_internal/common_nn.py:4921]: nice, I was going to ask for these dict entries but you beat me to it!

File: aten/src/ATen/native/LossNLL.cpp [hunk 14]
```diff
@@ -497,12 +507,67 @@ Tensor cross_entropy_loss_prob_target(
...
+    if (weight.defined()) {
+      // Expand weight to the correct number of dims for broadcasting with input / target
+      auto weight_broadcast_shape = SmallBuffer<int64_t, 5>(input.dim());
+      std::fill(weight_broadcast_shape.begin(), weight_broadcast_shape.end(), 1);
+      weight_broadcast_shape[1] = weight.size(0);
+      Tensor weight_ = weight.view(weight_broadcast_shape);
+
+      smooth_loss = -at::sum(input * weight_, 1);
+    } else {
+      smooth_loss = -at::sum(input, 1);
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:534]: nit: `-input.sum(1)`

**Review by jbschlosser — COMMENTED** [review]
Looking great!
* Implementation for class prob target looks solid
* Implementation for class indices target uses `nll_loss_nd` correctly AFAICT; it matches the current mean behavior of dividing by the sum of the target weights handles ignore_index
* Testing looks good - tests for consistency across index and prob targets and comparisons with manual references are present for all reductions. Tests for errors are present as well
* Formula for label smoothing looks correct throughout to me
* Docs for new arg in the module form look good, but need the same for for the functional form
* Note that to fully finish out the issue, we should probably add the new arg to the C++ module and functional forms as well. To get this in quickly, I'm okay with bumping this part to a separate PR

Some minor comments below, but nothing too crazy. Thanks for all your hard work in getting this in!

File: aten/src/ATen/native/LossNLL.cpp (hunk 6 above)
Comment (thomasjpfan) [reply aten/src/ATen/native/LossNLL.cpp:540]: I think we still need the `if (ignore_index >= 0)` to avoid the operation for the the default, `ignore_index=-100`, case?

Comment (jbschlosser) [reply aten/src/ATen/native/LossNLL.cpp:540]: Hm, I think the kernels for `NLLLoss` just do an unconditional comparison, but I might be wrong.

**Comment by jbschlosser** [comment]
@thomasjpfan FYI failing checks look related to C++ / python API disparity

File: aten/src/ATen/native/native_functions.yaml [hunk 15]
```diff
@@ -6652,7 +6652,7 @@
   device_check: NoCheck   # TensorIterator
   variants: method
 
-- func: cross_entropy_loss(Tensor self, Tensor target, Tensor? weight=None, int reduction=Mean, int ignore_index=-100) -> Tensor
+- func: cross_entropy_loss(Tensor self, Tensor target, Tensor? weight=None, int reduction=Mean, int ignore_index=-100, float label_smoothing=0.0) -> Tensor
```
Comment (jbschlosser) [inline aten/src/ATen/native/native_functions.yaml:6655]: FYI I think it's likely we'll need to keep the old signature around as well for FC reasons.

**Review by jbschlosser — COMMENTED** [review]
Getting pretty close to wrapping this up! Few more things I see needed:
* Update override for `cross_entropy` in `torch/overrides.py` with the new kwarg
* Update `torch/nn/functional.pyi.in` with the new kwarg for mypy
* C++ tests in `test/cpp/api/modules.cpp` and `test/cpp/api/functional.cpp`

**Review by jbschlosser — APPROVED** [review]
LGTM!

File: torch/onnx/symbolic_opset12.py [hunk 16]
```diff
@@ -65,14 +65,17 @@ def nll_loss_nd(g, self, target, weight, reduction, ignore_index):
...
-def cross_entropy_loss(g, self, target, weight, reduction, ignore_index):
+def cross_entropy_loss(g, self, target, weight, reduction, ignore_index, label_smoothing):
     # none reduction : onnx::Constant[value={0}]
     # mean reduction : onnx::Constant[value={1}]
     # sum reduction : onnx::Constant[value={2}]
     reduction = sym_help._maybe_get_const(reduction, "i")
     reduction_vals = ["none", "mean", "sum"]
     reduction = reduction_vals[reduction]
 
+    if label_smoothing > 0.0:
```
[the remark this answers is not in the document]
Comment (garymm) [reply torch/onnx/symbolic_opset12.py:77]: I took a look. Ideally we'd try to support this in ONNX, but this is OK for now.
